File: Java_version/LeetCode_Java/src/L380.java
import java.util.*;

public class L380 {
    List<Integer> valList = new ArrayList<>();
    Map<Integer, Integer> indexMap = new HashMap<>();
    Random rand = new Random();
    public L380() {

    }

    public boolean insert(int val) {
        if (!indexMap.containsKey(val)){
            valList.add(val);
            indexMap.put(val, valList.size() - 1);
            return true;
        }
        return false;
    }

    public boolean remove(int val) {
        if (indexMap.containsKey(val)){
            int temp = valList.get(valList.size() - 1);
            int index = indexMap.get(val);

            valList.set(index, temp);
            indexMap.put(temp, index);

            valList.remove(valList.size() - 1);
            indexMap.remove(val);

            return true;
        }
        return false;
    }

    public int getRandom() {
        return valList.get(rand.nextInt(valList.size()));
    }
}

File: L380.cpp
//
// Created by DuLi on 2023/10/23.
//

#include "L380.h"
#include <iostream>
#include <unordered_map>
#include<cstdlib>

class RandomizedSet {
private:
    std::vector<int> val_list;
    std::unordered_map<int, int> index_map;

public:
    /** Initialize your data structure here. */
    RandomizedSet() {
    }

    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    bool insert(int val) {
        if (index_map.find(val) == index_map.end()) {
            val_list.push_back(val);
            index_map[val] = val_list.size() - 1;
            return true;
        }
        return false;
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    bool remove(int val) {
        if (index_map.find(val) != index_map.end()) {
            int temp = val_list.back();
            int index = index_map[val];

            val_list[index] = temp;
            index_map[temp] = index;

            val_list.pop_back();
            index_map.erase(val);

            return true;
        }
        return false;
    }

    /** Get a random element from the set. */
    int getRandom() {
        int index = rand() % val_list.size();
        return val_list[index];
    }
};